src/jupyterlabcontroller/services/lab.py

```diff
@@ -69,175 +206,339 @@ async def create_lab(
             invalid.
         LabExistsError
             Raised if this user already has a lab.
+        OperationConflictError
+            Raised if some other operation (either spawn or delete) is already
+            in progress on this user's lab.
         """
-        selection = lab.options.image_list or lab.options.image_dropdown
+        username = user.username
+
+        # If the user was not previously seen, set up their data structure and
+        # monitor class.
+        if username not in self._labs:
+            monitor = _LabMonitor(
+                username=username,
+                timeout=self._config.spawn_timeout,
+                slack_client=self._slack,
+                logger=self._logger,
+            )
+            self._labs[username] = _State(monitor=monitor)
+
+        # Determine the image to use for the lab.
+        selection = spec.options.image_list or spec.options.image_dropdown
         if selection:
             reference = DockerReference.from_str(selection)
             image = await self._image_service.image_for_reference(reference)
-        elif lab.options.image_class:
-            image_class = lab.options.image_class
+        elif spec.options.image_class:
+            image_class = spec.options.image_class
             image = self._image_service.image_for_class(image_class)
-        elif lab.options.image_tag:
-            tag = lab.options.image_tag
+        elif spec.options.image_tag:
+            tag = spec.options.image_tag
             image = await self._image_service.image_for_tag_name(tag)
 
         # Check to see if the lab already exists. If so, but it is in a failed
-        # state, delete it first.
-        status = await self._lab_state.get_lab_status(user.username)
-        delete_first = status == LabStatus.FAILED
+        # state, we will delete the previous lab first.
+        state = await self.get_lab_state(user.username)
+        delete_first = bool(state and state.status == LabStatus.FAILED)
```

Comment:
You added a TERMINATED state; do we want to check that here as well?

Reply:
This code doesn't handle labs in a completed state yet. I'm going to tackle that in a separate diff since I think it may require rsp-restspawner changes as well. It just lays a bit of the groundwork for that.